### PR TITLE
chore(michelson encoder): improved Errors in the Michelson Encoder Package

### LINKED ISSUE #1995: Improve error in the Michelson-Encoder package
_Follow up of https://github.com/ecadlabs/taquito/issues/160_ as described in https://hackmd.io/@tca0T8LZT3au8LG6hSX20g/r1zTRTlfj_

Task 4: @taquito/michelson-encoder package - We use the same error if an issue happens in the Encode and Execute methods. In some situations, it is hard to identify what went wrong. Make sure we clearly distinguish if it is a bad input from the user or an issue from the Execute method.

### PR DESCRIPTION
**Pull request by AlirezaHaghshenas**
Refactored errors in the Michelson Encoder package to have better messages and be more informative to the user

closes #1995

Thank you for your contribution to Taquito.

Before submitting this PR, please make sure:

- [x] Your code builds cleanly without any errors or warnings
- [x] You have run the linter against the changes
- [x] You have added unit tests (if relevant/appropriate)
- [x] You have added integration tests (if relevant/appropriate)
- [x] All public methods or types have TypeDoc coverage with a complete description, and ideally an @example
- [ ] You have added or updated corresponding documentation
- [ ] If relevant, you have written a first draft summary describing the change for inclusion in Release Notes. 

In this PR, please also make sure: 

- [x] You have linked this PR to the issue by putting `closes #TICKETNUMBER` in the description box (when applicable)
- [ ] You have added a concise description on your changes
## Release Note Draft Snippet

__If relevant, please write a summary of your change that will be suitable for
inclusion in the Release Notes for the next Taquito release.__

### CODE REVIEW

File: packages/taquito-michelson-encoder/src/michelson-map.ts [hunk 1]
```diff
@@ -19,21 +20,52 @@ const michelsonMapTypeSymbol = Symbol.for('taquito-michelson-map-type-symbol');
...
-const isMapType = (
-  value: MichelsonV1Expression
-): value is { prim: 'map' | 'big_map'; args: [MichelsonV1Expression, MichelsonV1Expression] } => {
-  return 'args' in value && Array.isArray(value.args) && value.args.length === 2;
-};
+/**
+ *
+ * @throws {@link InvalidMapTypeError} when the argument passed to mapType is not a valid map type
+ */
+function validateMapType(value: MichelsonV1Expression): asserts value is {
```
Comment (hui-an-yang) [inline packages/taquito-michelson-encoder/src/michelson-map.ts:27]: great work!!

File: packages/taquito-michelson-encoder/src/schema/parameter.ts [hunk 2]
```diff
@@ -12,26 +12,55 @@ import { InvalidRpcResponseError, ParameterEncodingError } from './error';
...
+        val,
+        `The response.script.code has an element of type {prim: "parameter"}, but its args is not an array`
+      );
     }
-
     return new ParameterSchema(parameter.args[0]);
   }
 
+  /**
+   * @description Check if the Contract parameter is multiple entry point or not
```
Comment (dsawali) [inline packages/taquito-michelson-encoder/src/schema/parameter.ts:52]: thanks for adding all the descriptions!

File: packages/taquito-michelson-encoder/src/tokens/token.ts [hunk 3]
```diff
@@ -1,14 +1,15 @@
...
 import { TokenSchema } from '../schema/types';
+import { TaquitoError } from '@taquito/core';
 
 /**
  *  @category Error
  *  @description Error that indicates a failure when encoding invalid or incorrect data (e.g. if an address is expected but a number is received)
  */
-export abstract class TokenValidationError extends Error {
+export abstract class TokenValidationError extends TaquitoError {
   name = 'ValidationError';
```
Comment (hui-an-yang) [inline packages/taquito-michelson-encoder/src/tokens/token.ts:10]: might want to correspond the name declaration with the class name

File: packages/taquito-michelson-encoder/src/michelson-map.ts [hunk 4]
```diff
@@ -19,21 +21,52 @@ const michelsonMapTypeSymbol = Symbol.for('taquito-michelson-map-type-symbol');
...
 
-  constructor(public readonly value: any, public readonly type: any, errorType: 'key' | 'value') {
-    super(`${errorType} not compliant with underlying michelson type`);
+  constructor(
+    public readonly value: any,
+    public readonly type: any,
+    objectType: 'key' | 'value',
+    public readonly reason: any
+  ) {
+    super(
```
Comment (hui-an-yang) [inline packages/taquito-michelson-encoder/src/michelson-map.ts:62]: might still have some super(message) in places, hope you don't mind updating them

File: packages/taquito-michelson-encoder/src/schema/error.ts
Comment (hui-an-yang) [inline packages/taquito-michelson-encoder/src/schema/error.ts:1]: would be great to rename file to errors.ts

File: packages/taquito-michelson-encoder/src/michelson-map.ts [hunk 5]
```diff
@@ -1,15 +1,16 @@
...
  *  @category Error
  *  @description Error that indicates an invalid map type being passed or used
  */
-export class InvalidMapTypeError extends Error {
-  public name = 'InvalidMapTypeError';
-  constructor(public mapType: string) {
-    super(`The map type '${mapType}' is invalid`);
+export class InvalidMapTypeError extends TaquitoError {
+  constructor(public readonly mapType: any, public readonly reason: string) {
+    super(`The map type '${JSON.stringify(mapType)}' is invalid. Reason: ${reason}.`);
```
[the remark this answers is not in the document]
Comment (AlirezaHaghshenas) [reply packages/taquito-michelson-encoder/src/michelson-map.ts]: Done

File: packages/taquito-michelson-encoder/src/tokens/token.ts (hunk 3 above)
Comment (AlirezaHaghshenas) [reply packages/taquito-michelson-encoder/src/tokens/token.ts]: Good catch, fixed

File: packages/taquito-michelson-encoder/src/schema/error.ts
Comment (AlirezaHaghshenas) [reply packages/taquito-michelson-encoder/src/schema/error.ts:1]: Done

File: packages/taquito-michelson-encoder/src/schema/errors.ts [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    message += `Script: ${JSON.stringify(script)}`;
+    this.message = message;
+  }
+}
+
+/**
+ *  @category Error
+ *  @description Error that indicates an invalid RPC response being passed or used
+ */
+export class InvalidRpcResponseError extends TaquitoError {
```
Comment (dsawali) [inline packages/taquito-michelson-encoder/src/schema/errors.ts:47]: is there an already existing error that's somewhat similar? might be able to reuse that

**Review by dsawali — APPROVED** [review]

**Pull request merged**